### PR TITLE
Fix entity gui and some assorted stuff

### PR DESCRIPTION
**Pull request by liach**
**The Issue**
 - Wrong GUI code might have broken tnt carts.
 - #1718 
 - Advancements aren't trigger for strucutre logics.
 - #1567
 
**The Proposal**
 - Add hooks for structure logic
 - Fix `-1`
 - Write capacity in packet sync
 - fix a minor issue in depth detection 

**Possible Side Effects**
 - Haven't seen any yet.
 
**Alternatives**
 - Just tried to do minimal change for these bugs.

### CODE REVIEW

File: src/main/java/mods/railcraft/common/blocks/multi/TankDefinition.java [hunk 1]
```diff
@@ -61,7 +61,7 @@ public void defineRecipes(Block block) {
             Crafters.blastFurnace().newRecipe(block)
                     .name("railcraft:smelt_iron_tank")
                     .time(IBlastFurnaceCrafter.SMELT_TIME * 4)
-                    .output(RailcraftItems.INGOT.getStack(4, Metal.STEEL)).slagOutput(4)
+                    .output(RailcraftItems.NUGGET.getStack(4, Metal.STEEL)).slagOutput(1)
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/multi/TankDefinition.java:64]: Why? Didn't I just revert this change? Almost all the other smelting recipes are 1:1 on the iron:steel ratio.

**Review by CovertJaguar — CHANGES_REQUESTED** [review]

File: src/main/java/mods/railcraft/common/blocks/multi/TankDefinition.java (hunk 1 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/blocks/multi/TankDefinition.java:64]: this is for #1718. Four iron plates can craft 8 tanks walls.

Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/blocks/multi/TankDefinition.java:64]: Gotcha!

**Review by CovertJaguar — APPROVED** [review]
